### PR TITLE
Initial implementation of a search API.

### PR DESCRIPTION
**Pull request by EreMaijala**
This is finally the long-promised proposed implementation of a search/record API for VuFind. A very similar API has been in production use in Finna for a while now, and for now there haven't been many issues with it.

The following design criteria were followed when the API endpoints were designed:
- Easy to use with modern tools (JSON, CORS, Swagger specification).
- Parameters as similar as possible to the search functionality in the VuFind UI so that a search can made and filtered etc. and then parameters can then be easily moved to the API.
- Extensible with local fields etc. as necessary.
- Take advantage of record processing available in the record drivers to provide data that's ready for re-use.
- Include version in the end points from the beginning so that new versions with possible incompatibilities can be created without affecting existing API users.
- Use permissions so that e.g. IP address filters can be used to limit access to the API.
- Others I can't remember right now :)

This implementation contains groundlaying work for APIs and tries to take into account other endpoints too. That's why some of the common functionality is in a trait that can be shared between API controllers that extend other controllers.

There's currently no support for authorization or rate limits. We decided to provide the search functionality without limits, but if e.g. holdings information retrieval or other functionality affecting third parties is added, these will become relevant.

I haven't verified that a fully functional client can be created using the Swagger spec, but at least it looks ok in the included Swagger UI.

I would appreciate initial reactions before the VuFind Summit, but there's no rush to get this merged or anything before that.

TODO:
- [x] Wiki page
- [x] Investigate possible issues with fetching records with special characters in the IDs
- [x] Controller cleanup?
  - [x] Fix hierarchical facet translation broken by refactoring (use view helper instead of raw translator?)
- [x] Create tests for new Formatter classes.
  - [x] FacetFormatter
  - [x] RecordFormatter

### CODE REVIEW

**Comment by EreMaijala** [comment]
Forgot to mention: Go to http://vufind-address/api for the Swagger UI.

File: module/VuFind/config/module.config.php [hunk 1]
```diff
@@ -830,6 +831,41 @@
 $routeGenerator->addDynamicRoutes($config, $dynamicRoutes);
 $routeGenerator->addStaticRoutes($config, $staticRoutes);
 
+// API routes
+$config['router']['routes']['searchApiHome'] = [
```
Comment (demiankatz) [inline module/VuFind/config/module.config.php]: Perhaps it would make sense at some point to add an "addApiRoute" method to the route generator to reduce the verbosity of this configuration, especially if we expect over time to be adding a lot of variations with different version numbers.

File: module/VuFind/src/VuFind/Controller/ApiInterface.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Additional functionality for API controllers.
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:controllers Wiki
+ */
+interface ApiInterface
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/ApiInterface.php]: Do we really need this interface? If its only purpose is to define some constants, it seems like we could just as easily move them to the Trait. Of course, this could be justified if you anticipate needing it for marker purposes or to allow alternate implementations of API functionality... It just seems a little odd as currently implemented.

File: module/VuFind/src/VuFind/Controller/ApiTrait.php [hunk 3]
```diff
@@ -0,0 +1,732 @@
...
+     * Available record fields.
+     *
+     * Key is the field name that can be requested. Value is an array containing the
+     * method name to call (either in this class or the record driver) and Swagger
+     * specification fields describing the returned data.
+     *
+     * @var array
+     * @see http://swagger.io/specification/
+     */
+    protected $recordFields = [
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/ApiTrait.php]: This seems like a strong candidate for something that could be moved to a YAML file. Rather than including this data through a trait, it might make more sense to load it from a config file and inject it into the controllers through a factory. You could use a cache (as we do with searchspecs) to avoid the performance disadvantages of YAML parsing. I think the configuration file approach has a significant advantage since a major weakness of Traits (from the perspective of an application that encourages separation of local customizations) is that you can't modify them without changing core code, and the only way to override them is to extend all of the classes that utilize them. It's best not to put "variable" things in Traits for this reason.

File: module/VuFind/src/VuFind/Controller/ApiTrait.php [hunk 4]
```diff
@@ -0,0 +1,732 @@
...
+                . urlencode("$base/api?swagger");
+            return $this->redirect()->toUrl($url);
+        }
+        $response = $this->getResponse();
+        $headers = $response->getHeaders();
+        $headers->addHeaderLine('Content-type', 'application/json');
+        $config = $this->getConfig();
+        $results = $this->getResultsManager()->get($this->searchClassId);
+        $options = $results->getOptions();
+        $searchConfig = $this->getConfig($options->getSearchIni());
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/ApiTrait.php]: Is there a compelling reason to read raw data out of the .ini files rather than making use of the processed data from the Options object? There are some cases where the PHP code may modify or extend the configuration settings, and those options will be lost with this approach. (This may or may not be important for the purposes of the API -- but it seems safer/more consistent to use the more abstract interfaces rather than raw data whenever practical).

File: module/VuFind/src/VuFind/Controller/SearchApiController.php [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+
+        return $this->output($response, self::STATUS_OK);
+    }
+
+    /**
+     * Search action
+     *
+     * @return \Zend\Http\Response
+     */
+    public function searchAction()
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/SearchApiController.php]: This is an extremely long method; I suspect readability could be improved by the introduction of support methods/classes.

Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/SearchApiController.php]: ...in general, I wonder if we should introduce a \VuFind\API namespace and put a bunch of helper classes there to make the controllers thinner. Obviously it makes sense for the controllers to handle the high-level framework stuff of constructing responses and parsing input, but anything related to formatting API responses, etc., might be better positioned in its own classes (which would enable better testing, greater separation from framework-specific details, etc., etc.)

File: public/swagger-ui/css/print.css [hunk 6]
```diff
@@ -0,0 +1,1367 @@
+/* Original style from softwaremaniacs.org (c) Ivan Sagalaev <Maniac@SoftwareManiacs.Org> */
```
Comment (demiankatz) [inline public/swagger-ui/css/print.css]: How does all of this swagger-ui stuff fit in? Is this just an existing open source project dropped into a directory? (That's what it looks like at a glance). Is it essential to the API implementation, or just a convenient way of testing it? Assuming it is a "vendor" type of dependency, would it make sense, rather than including it in the PR, to add an appropriate .gitignore rule and include some kind of task to automatically install it on demand (using grunt or Composer or Phing, depending on which options are most feasible)? As it is, it seems to be introducing a fair amount of noise to this PR.

Comment (demiankatz) [inline public/swagger-ui/css/print.css]: Also, it is important to have an easy way of upgrading the dependency in the future as new versions are released -- the more automated we can make that procedure, the better.

File: module/VuFind/src/VuFind/Controller/SearchApiController.php (hunk 5 above)
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/SearchApiController.php]: Ok, I'll work on that.

File: module/VuFind/config/module.config.php (hunk 1 above)
Comment (EreMaijala) [inline module/VuFind/config/module.config.php]: True, I'll do that.

File: module/VuFind/src/VuFind/Controller/ApiInterface.php (hunk 2 above)
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/ApiInterface.php]: Traits can't define constants, but if we move the stuff in the trait to separate class, that issue no longer exists.

File: module/VuFind/src/VuFind/Controller/ApiTrait.php (hunk 4 above)
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/ApiTrait.php]: No, I just failed to realize I can get them from the Options object. Will change.

**Comment by demiankatz** [comment]
One other possible thought on code organization: perhaps it would make sense to move all of this into its own ZF module, similar to VuFindAdmin and VuFindDevTools. That's a good way to neatly bundle together configurations and code, and to make it very easy to disable without any possibility of unwanted access slipping through the cracks should anyone have concerns about it.

**Comment by EreMaijala** [comment]
That's indeed a great idea worth checking out. I'll work on that.

**Comment by EreMaijala** [comment]
Do you have a preference where to put the helper classes in the VuFindApi module? src/VuFindApi/Helper?

File: module/VuFind/src/VuFind/Controller/ApiTrait.php (hunk 4 above)
Comment (demiankatz) [inline module/VuFind/src/VuFind/Controller/ApiTrait.php]: It's possible that for some things you may need to add accessors... but I think that's worth doing if it's not already done.

File: public/swagger-ui/css/print.css (hunk 6 above)
Comment (demiankatz) [inline public/swagger-ui/css/print.css]: I suspect that this could be handled with a Grunt task. @crhallberg might be able to provide some advice on that if you're not already familiar with the technology.

**Comment by demiankatz** [comment]
I think VuFindApi\Helper could make work, though it also may depend on how the tasks break down. I could imagine VuFindApi\Search and VuFindApi\Record might also make sense under some circumstances.

File: public/swagger-ui/css/print.css (hunk 6 above)
Comment (EreMaijala) [inline public/swagger-ui/css/print.css]: Ah, I didn't realize Grunt is required. That implies npm, so it shouldn't be a problem at all.

Comment (demiankatz) [inline public/swagger-ui/css/print.css]: Yes, we added an official VuFind Gruntfile to master quite recently (in time for release 3.1), so you should have an existing foundation to add tasks to.

**Comment by EreMaijala** [comment]
Second iteration now committed. I moved the API stuff into VuFindApi module and I believe I've now addressed review comments other than these:
- There's still a trait and an interface. The interface now serves a purpose, and I feel that the trait is much easier than a helper, but I can still change it if necessary.
- I haven't added an AddApiRoute helper as I'm not sure how the APIs will evolve and whether it will be useful (the routes might be so different that a helper wouldn't make it any simpler).

**Comment by demiankatz** [comment]
That all makes sense to me. We can always refactor from Trait to Helper if there is a reason to do so, and putting the configuration in the VuFindApi module gets the verbose route configuration "out of the way" and at least alleviates the need for an AddApiRoute helper -- though as you say, we can revisit it in future if we find ourselves creating a lot of similar route definitions.

I'll try to actually give this a test drive (and another, more thorough, code review) some time before the Summit, but it's looking like a very solid start. Thanks again!

File: config/vufind/httpd-vufind.conf [hunk 7]
```diff
@@ -79,8 +79,8 @@ Alias /vufind /usr/local/vufind/public
 
   # This line specifies additional Zend Framework 2 modules to load after the standard VuFind module.
   # Multiple modules may be specified separated by commas.  This mechanism can be used to override
-  # core VuFind functionality without modifying core code.
-  #SetEnv VUFIND_LOCAL_MODULES VuFindLocalTemplate
+  # core VuFind functionality without modifying core code or to enable the VuFindApi module.
+  #SetEnv VUFIND_LOCAL_MODULES VuFindApi,VuFindLocalTemplate
```
Comment (demiankatz) [inline config/vufind/httpd-vufind.conf]: This change will break the installer's ability to enable custom modules (see https://github.com/vufind-org/vufind/blob/master/install.php#L515). I'd be inclined to just enable the API module as part of the default list (https://github.com/vufind-org/vufind/blob/master/config/application.config.php#L5) and worry about coming up with a mechanism to disable core modules through configuration if the need arises. (This would be consistent with how VuFindAdmin is currently handled, and following that precedent probably makes life easier for most people).

File: module/VuFind/src/VuFind/Config/SearchSpecsReader.php [hunk 8]
```diff
@@ -37,119 +36,16 @@
...
-            foreach ($this->parseYaml($defaultParent) as $section => $contents) {
-                if (!isset($results[$section])) {
-                    $results[$section] = $contents;
-                }
-            }
-        }
-
-        return $results;
+        parent::__construct($cacheManager);
+        $this->cacheName = 'searchspecs';
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Config/SearchSpecsReader.php]: Is there actually a compelling need for the searchspecs to be cached separately from other YAML? Might it make sense to just rename the searchspecs cache to YAML and be done with it? The only advantage of separation I can think of is that it would allow different caches to be cleared independently, but that seems like a very minimal benefit.

It could still potentially make sense to have a SearchSpecsReader subclass of YamlReader since I could imagine we might eventually want to differentiate some other behavior (such as parent inheritance processing) but for now we might be able to get away with making this an empty class.

If we do go down that road, we should also ensure that any special handling of the 'searchspecs' cache (for example, in the Admin module's cache clearing functionality or in the cache manager itself) gets appropriately updated. I'm not sure if such special handling even exists, but it's worth checking for.

File: package.json [hunk 9]
```diff
@@ -14,9 +14,13 @@
...
     "grunt-less-to-sass": "0.0.10",
-    "jit-grunt": "^0.9.1"
+    "jit-grunt": "^0.9.1",
+    "swagger-ui": "^2.2.5"
   },
   "devDependencies": {
     "grunt-contrib-watch": "~0.6.1"
+  },
+  "scripts": {
+    "postinstall": "rm -rf public/swagger-ui; cp -r node_modules/swagger-ui/dist public/swagger-ui; sed -i .orig \"s/defaultModelRendering: 'schema'/defaultModelRendering: 'model'/\" public/swagger-ui/index.html; sed -i .orig 's/url = \".*\"/url = \"..\\/..\\/api\\/v1\\?swagger\"/' public/swagger-ui/index.html"
```
Comment (demiankatz) [inline package.json]: We should update the "package" task in build.xml to make the appropriate calls to install this stuff and to ensure that the node_modules directory or other non-essential Node.js-specific code stays out of the distribution packages at release time. I can help with this if you're not already familiar with Phing -- just put a TODO item in the top-level description and let me know and I'll take care of it.

**Comment by demiankatz** [comment]
I'm having trouble with the swagger-ui installation. When I run `npm install` it complains that the postinstall script is failing. When I manually run `rm -rf public/swagger-ui; cp -r node_modules/swagger-ui/dist public/swagger-ui; sed -i .orig "s/defaultModelRendering: 'schema'/defaultModelRendering: 'model'/" public/swagger-ui/index.html; sed -i .orig 's/url = ".*"/url = "..\/..\/api\/v1\?swagger"/' public/swagger-ui/index.html`, I get this output:

<pre>
sed: -e expression #1, char 1: unknown command: `.'
sed: -e expression #1, char 1: unknown command: `.'
</pre>

Is this a platform-specific issue (I'm testing on Ubuntu 16), am I running the wrong command, or is there a problem in the configuration?

**Comment by demiankatz** [comment]
Also, a side note: we should probably add public/swagger-ui to the .gitignore file.

**Comment by demiankatz** [comment]
In spite of the postinstall script failure, I was able to get this working well enough to explore a little bit. I'm quite impressed! Just one more question:

Right now I assume that this is more or less hard-coded to wrap around the biblio index in Solr. That alone gives us a lot of power and I certainly don't want to create excessive scope-creep... but should we account for the possibility of supporting additional search backends in the design somehow? While I can't imagine people would want to use the VuFind API to access things that have their own APIs (that's inefficient as well as possibly in violation of some terms of service), I can at least imagine a desire to access other Solr cores (website, authority, something locally customized). Do we want to somehow reflect VuFind's search backend system either through the URL or an API parameter? If so, how should we manage separate config files for separate cores? (Apologies if this is already being accounted for and I just missed it -- I haven't done a deep dive into the code yet).

In any case, like I say, I don't expect that we have to implement any of this for round 1, but I just want to be sure the initial design allows room for us to grow in this direction if the need arises.

File: public/swagger-ui/css/print.css (hunk 6 above)
Comment (crhallberg) [inline public/swagger-ui/css/print.css]: Looking at the swagger-ui docs, Grunt is not required to get it working, only for developing swagger itself. I did have to sudo to run `npm install` successfully, though.

File: config/vufind/httpd-vufind.conf (hunk 7 above)
Comment (EreMaijala) [inline config/vufind/httpd-vufind.conf]: I didn't realize that. Now fixed.

File: package.json (hunk 9 above)
Comment (EreMaijala) [inline package.json]: You should be able to install the swagger-ui easily without packaging the whole thing. Needing grunt for that is quite heavy, but if you think it's better, I could add a separate grunt task and make it part of the package task.

**Comment by EreMaijala** [comment]
The sed issue should be fixed now, .gitignore updated and VuFindApi module enabled by default.

File: package.json (hunk 9 above)
Comment (demiankatz) [inline package.json]: We talked about this a little in person yesterday, but I think we may not be talking about the same thing. Just in case clarification is needed: I rely on [this task](https://github.com/vufind-org/vufind/blob/master/build.xml#L295) to create VuFind releases (`phing package` from the command line). Since we now have an additional step to get all of the files in place for distributing VuFind, I need to be sure when I run the package task that npm has done its job. It doesn't matter from this perspective whether we use grunt or npm -- we just need to be sure that something puts the files where they belong before we zip them up for distribution.

Comment (EreMaijala) [inline package.json]: Aha, I didn't realize "npm install" isn't part of the process. I'll see if we can actually install swagger-ui with composer.

**Comment by EreMaijala** [comment]
Swagger-ui is now installed by composer. That should solve the packaging issue.

**Comment by EreMaijala** [comment]
Initial wiki pages are now available at https://vufind.org/wiki/development:apis

**Comment by demiankatz** [comment]
@EreMaijala, thanks again for all of the progress on this. What are the remaining outstanding issues? Off the top of my head, I can think of:

- There may still be some refactoring to do to reduce controller size and create helpers.
- My question about supporting non-default search backends should probably be answered (like I say, we don't necessarily need to implement it yet, but we should at least make sure the API design has a logical place where a backend could be specified)
- New topic: one small disadvantage to the composer.json post-install hook is that it's Linux-specific. Prior to that change, Composer worked the same in Windows and Linux. That will no longer be the case. Probably not a big deal -- I don't think a lot of people are going to be developing VuFind under Windows without access to a Linux box, and things like #769 are going to add similar Windows-limiting factors -- but it seemed at least worth mentioning for the sake of discussion.

Should we add any of these things to the TODO list at the top? Am I forgetting anything else?

**Comment by EreMaijala** [comment]
I thought I already answered your question about other backends, but apparently I failed to submit it or something. Anyway, I've thought about it, but since there are different fields and features available in the backends, they should have their own endpoints. I believe adding them will be trivial when we have the basic structure in place. There are probably methods that can be shared in a trait, but I didn't do that yet since I wasn't sure if another trait will be needed or if they're generic enough to go to ApiControllerTrait.

I'm open to suggestions on how to further clean up the controller.

I'll need investigate one more issue we noticed with fetching records with special characters in their IDs. I'll make it a TODO list item.

**Comment by EreMaijala** [comment]
The special character issue was my bad. There was a plus sign in the record ID and while it's not encoded in the normal record URL where it's part of the path, it obviously needs to be encoded in the Record API since it's a query parameter.

**Comment by demiankatz** [comment]
I think the main question with the backend issue is whether we need to adjust the current URL structure to make room for future growth. Right now we have api/v1/search and api/v1/record. Do we want api/v1/solr/(search|record)? Do we want api/solr/v1/(search|record)? Do we want to work on the assumption that v2 may rework this? Do we want to work on the assumption that "search" and "record" will always map to Solr, and other endpoints will have different names following some pattern?

Honestly I have no strong opinions about any of this -- just throwing some ideas out for consideration. I think it's mainly a matter of balancing convenient, readable URLs with avoiding having to change legacy URLs at some point in the future.

**Comment by EreMaijala** [comment]
Very good questions, and I have no definitive answers either except for one thing: let's not put solr in the URL. I think it's better to think about the logical indexes instead: biblio, authority, reserves, website. My initial idea was to have biblio as the primary without a separate endpoint, but it would be justifiable to have it in the URL structure too. I might be biased since we don't currently have anything but biblio.

I think it may become quite muddy if we have different API versions for different indexes, so I'd say make it api/v1/... It's anyway necessary to increase the API version if any of the endpoints become incompatible with the previous one (and even in that case keep the old endpoint for compatibility). Other than that, I think it's ok to keep the API URL the same if we just fix bugs or add new stuff in a way that doesn't break existing functionality.

To sum it up, I'd be fine with either of these:

/api/v1/(search|record)
/api/v1/biblio/(search|record)

**Comment by demiankatz** [comment]
It occurs to me that one argument in favor of keeping it simple with /api/v1/(search|record) is that our current controller architecture has SearchController and RecordController for the equivalent user-facing routes, so the URL structure would be consistent between human-readable and machine-readable information. Including "biblio" in the route is of course more explicit, but it's also not really consistent with anything else that VuFind is currently doing.

Let's talk about this in a little more detail on the next dev call, and in the meantime, I'll try to find time to do another close code review, but I think we're getting close to being able to merge this so people can start playing with it!

**Comment by demiankatz** [comment]
@EreMaijala, I just took another look at this, and I still think it would be valuable to refactor SearchApiController to be less huge and more testable. One proposal:

Create an AbstractFormatter class containing the filterArrayValues and resetArrayIndices methods. Create subclasses of this called FacetFormatter and RecordFormatter containing the relevant formatting methods. Create SearchApiController::getFacetFormatter and ::getRecordFormatter to create extension points for overriding these classes as needed. Then code like this:

<pre>
        $records = [];
        foreach ($results->getResults() as $result) {
            $records[] = $this->getFields($result, $requestedFields);
        }

        $this->filterArrayValues($records);

        /* ... snip ... */

        if ($records) {
            $response['records'] = $records;
        }
</pre>

Could become:

<pre>
        if ($records = $this->getRecordFormatter()->format($results->getResults()) {
            $response['records'] = $records;
        }
</pre>

(And obviously the rest of that logic moves to the format method). I think this also reduces some redundancy between the record and search methods.

Anyway, that's just a sketch of an idea based on a cursory examination... maybe it doesn't actually break down this tidily. But let me know if you think this sounds like a reasonable idea. I might be able to find time to help with this final bit of refactoring if you think my proposed approach makes sense but you don't have time to work on it yourself. (I can also help with a different approach if you have a counter-proposal).

**Comment by EreMaijala** [comment]
Probably a good idea. I could use some help, though, so feel free to work on it if time permits. Thanks!

**Comment by demiankatz** [comment]
I may have some time for this later in the week. If you decide to work on this in the meantime, just ping me on Slack to be sure we're not doing the same thing at the same time.

**Comment by demiankatz** [comment]
@EreMaijala, I took a first stab at refactoring to helpers in commit https://github.com/vufind-org/vufind/pull/819/commits/343ffb41ce95aeff94add652091e8d0c324545cd. It's entirely possible that I've broken something, as this required some significant adjustments, though I've at least confirmed that for some simple test requests, the results are identical before and after the refactoring.

Obviously, we ought to write some tests before we merge this. If you like the general direction that I've gone in here, I might be able to find time to write some tests next week... but I don't want to spend time on test development until we're in agreement about the general shape of the code.

One design question: I notice that there is some logic which first checks for a method in the controller (now, following the refactoring, the RecordFormatter) and then fails over to the record driver. I wonder if this might potentially lead to some unwanted name collisions -- i.e. what if the user wants to access a record driver method that happens to have the same name as a formatter method? Should we do some kind of namespacing/prefixing to reduce the risk of collisions (and/or unwanted arbitrary code executions)? Or am I missing some existing protection against this that I simply didn't see in my cursory code read?

**Comment by EreMaijala** [comment]
I like what you've done, apart from breaking translation of hierarchical facets by changing the translate view plugin to VuFind\Translator. If you think we shouldn't use the plugin here, the translation with the handling of TranslatableString needs to be refactored out of the plugin.

I did a couple of changes to fix the Swagger spec creation, but didn't touch the translation yet.

There's currently no mechanism to avoid collisions other than naming the methods in the formatter so that it's unlikely. I'd be happy with a namespace prefix or something too (maybe just Formatter::getSomething for formatter methods?).

In summary I think we can continue on what you've done, and I'd appreciate help creating the tests.

**Comment by demiankatz** [comment]
I have no problem with changing the translation back to using the view helper. When I started refactoring, I didn't realize that there would be a dependency on the view helper manager, so I tried to avoid using view helpers... but as the work progressed, I ended up injecting the helper manager, so probably the easiest thing to do is to remove the external translator dependency entirely, and go back to pulling the translator out of the helper manager when needed. I'll do this if I get to it first, but feel free to take care of it if you have time before I do. I've created a TODO checkbox so we can keep track of the task (I've also created a TODO for tests, which I'll try to start on this week).

I like the idea of the Formatter prefix -- we could either require a prefix, or else we could use it as an optional means of disambiguation; whichever you prefer. I'd also suggest that, at very least, perhaps we should filter dynamic calls to the Formatter to exclude any method that does not start with the word "get" (just to avoid people from setting up weird and problematic scenarios).

In any case, thanks for taking a look -- I'll leave another comment when I've had time to make further progress.

**Comment by EreMaijala** [comment]
I think it's clearest if the Formatter prefix is required for all formatter methods, and the "get" requirement is a sound idea. So in practice `method: getUniqueID` would mean record driver's method, and `method: "Formatter::getUniqueID"` would mean RecordFormatter's getUniqueID. Then we can get rid of the method_exists check too.

**Comment by demiankatz** [comment]
Sounds good to me!

**Comment by demiankatz** [comment]
Thanks, @EreMaijala, this is looking great. I still need to find time for test-writing, but hopefully I can address that later this week. We should probably leave this open at least until the next dev call, but if tests get written between now and then and there are no new problems or objections, I think we may be in good shape to merge shortly thereafter.

**Comment by demiankatz** [comment]
@EreMaijala, I have begun some work on writing a FacetFormatterTest, but I've encountered some issues that might need to be addressed in the core code:

1.) The UrlQueryHelper HTML-encodes URLs by default. This results in `&amp;` entities in URLs when multiple filters are applied. This seems undesirable in the JSON API. Is this intentional, or is this a problem? If it's a problem, we may need to add some new parameters to the helper to adjust escaping. It may also be worth considering a significant redesign of UrlQueryHelper. I'm not particularly happy with the interface of that class, and I suspect we could do better -- the question is whether we can easily improve it without introducing huge BC breaks in the process.

2.) If a facet is applied, its URL includes its value twice, because FacetFormatter is calling addFacet() regardless of whether or not isApplied is true. I think we want to put a condition there and simply call getParams() when the facet is already applied.

I'm going to put test-writing on hold until we've had a chance to talk about these issues.

**Comment by demiankatz** [comment]
I've pushed up my work in progress on the test so we don't lose it... the bits that are currently going to cause failures are commented out until we sort out the details.

**Comment by EreMaijala** [comment]
1.) That's an oversight and not intentional. UrlQueryHelper is indeed quite terrible, but here we could manage with changes to make encoding & optional, or if it's difficult, reverse it after the fact..

2.) True. 

I'll try to get around to fixing these, but feel free to do it if you have a chance. :)

**Comment by demiankatz** [comment]
@EreMaijala, if I find time for this later in the week, I'll let you know. In the meantime, this situation has inspired me to open #850, which is a first stab at revising UrlQueryHelper to have a better interface. I think I've managed to do this in a backward-compatible way, but more eyes on the revisions would be much appreciated. The idea is that UrlQueryHelper now implements a __toString() method, and all of the methods that previously returned a string now return an instance of UrlQueryHelper. This means that operations can be stacked through a fluent interface, and the design now has more of an immutable feel to it (though it's not currently a strictly immutable implementation). This still has some rough edges, but I think it's a step in the right direction.

If you like the work in #850, we could consider polishing and merging it in advance of this PR so we can take advantage of it... but on the other hand, if you'd rather get this moving quickly, we can go ahead and add $escape parameters to a couple more methods and then roll that into the list of things supported as BC features by the new version of the code. Let me know what you think!

**Comment by demiankatz** [comment]
In the interest of moving forward, I've also solved the facet formatting problems in the simplest possible way here: https://github.com/vufind-org/vufind/pull/819/commits/8830804559ce603c6d59832d57b3f322701b7232 -- if we end up merging #850 quickly, we can back off the changes to UrlQueryHelper here and use the new interface instead.

**Comment by demiankatz** [comment]
@EreMaijala, I've set up tests for the basic facet formatting as well as the filtering capabilities of the formatter. I haven't done any tests specifically for hierarchical data yet, as I'm less familiar with those data structures. If you'd like to add another test to the suite, please feel free.

I'd still like to write a test for the RecordFormatter as well; I'll try to find time to start on that later in the week.

**Comment by EreMaijala** [comment]
Thanks, Demian. I'll expand the test to include hierarchical facets.

**Comment by EreMaijala** [comment]
I've now added a couple of hierarchical facet tests. They make the tests fail, though, because HierarchicalFacetHelper is also using UrlQueryHelper. There are so many methods that would have needed the $escape parameter that I decided to wait for the new UrlQueryHelper instead. Does that make sense?

**Comment by demiankatz** [comment]
Yes, that sounds reasonable. I'll make further progress on #850 a priority so we can get back to this soon.

**Comment by demiankatz** [comment]
@EreMaijala, I've merged #850 to master and merged master into this branch. I've managed to get the revised tests closer to passing, but something is still not quite right. Can you take a look when time permits?

**Comment by EreMaijala** [comment]
Thanks! I've made some adjustments and the tests and the actual API should be working properly now.

**Comment by demiankatz** [comment]
Great, thank you! I'll try to work on the remaining RecordFormatterTest soon so we can wrap this up.

File: module/VuFindApi/src/VuFindApi/Formatter/RecordFormatter.php [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+     */
+    protected function getSource($record)
+    {
+        if ($sources = $record->tryMethod('getSource')) {
+            $result = [];
+            $translator = $this->helperManager->get('translate');
+            foreach ($sources as $source) {
+                $result[] = [
+                    'value' => $source,
+                    'translated' => $translator->translate("source_$source")
```
Comment (demiankatz) [inline module/VuFindApi/src/VuFindApi/Formatter/RecordFormatter.php]: During test-writing, I noticed this; not sure what this is for exactly, and guessing it might be something related to your local practice (VuFind supports a getSource method, but it's hardly used anywhere). I don't object to adding this, but if we do, can we consider using a TextDomain (source::) instead of a prefix (source_)?

Comment (EreMaijala) [inline module/VuFindApi/src/VuFindApi/Formatter/RecordFormatter.php]: That's indeed local to us and predates the availability of text domains. I think it's safe to remove from the basic API for now since it doesn't return anything useful in the default Solr drivers.

**Comment by EreMaijala** [comment]
I've added some basic tests and a test for getRecordPage with a not-so-complicated mock helper since it's way easier to mock the helper than its dependencies. Do you find this acceptable?

**Comment by demiankatz** [comment]
Thanks for the additional tests -- I think this is a very solid start. I'll hold off on merging this until after we discuss it again on the next dev call... but if nobody there objects, I think we're ready to go. Once this has been merged and has gone through continuous integration, we can look at the coverage reports and perhaps add a few more test cases to bump the coverage higher.

**Comment by demiankatz** [comment]
No one has objected, so I've gone ahead and merged this. Once the next Jenkins build completes, we should look at test coverage for the new VuFindApi module and see if we need to make any improvements there.

**Comment by berkerY** [comment]
Hi @demiankatz , @EreMaijala 

is it possible to fetch data via vufind API from multiple sources (e.g. Solr + Summon) ? 
I think, current implementation of API delivers data only from Solr.

Bests,
Berker

**Comment by demiankatz** [comment]
@berkerY, you are correct that the API is presently Solr-oriented. We had some discussion about supporting other search backends, but the question arose of whether it would be appropriate to build an API that was essentially proxying another API. (There are certainly arguments in favor of this -- mainly to do with taking advantage of VuFind's record driver logic -- but on the other hand, it is probably not the most efficient approach, and I imagine there might be some terms of service implications in some cases).

What is your use case?
